README

BUC Customizer
==============

Description
-----------
Handy tool to customize guest session of public computers in the library of
Fermo (Italy). 

BUC Customizer is based on guest-session-prefs-lightdm by Gunnar Hjalmarsson.
Chances are you shouldn't install this software at all unless you exactely
know what you are doing and why this software has been created.

Authors, copyright and license
-----------------------------
- Authors: Gunnar Hjalmarsson, Andrea Colangelo
- Copyright (C) 2011-2012 Gunnar Hjalmarsson
- Copyright (C) 2012-2013 Andrea Colangelo
- BUC Customizer is distributed under the terms of GNU GPL3, see:
  http://www.gnu.org/licenses/gpl-3.0.txt
- shut.svg and shut_red.svg are Copyright (C) 2013 Marco Alici and are released
  under the terms of CC-BY-SA 3.0, see:
  http://creativecommons.org/licenses/by-sa/3.0/


debian/changelog
buc (5ubuntu0) UNRELEASED; urgency=medium

  * *.grotta: add to handle computers in Grottazzolina's library. 
  * debian/buc-common.post*: add support for xfce4-terminal.
  * shut.grotta.1: Add.
  * debian/control:
    + add a binary stanza for grotta.
    + properly declare Breaks among the three binary packages
    + bump Standards-Version, no changes required.
  * debian/grotta.install: add accordingly. 
  * debian/grotta.manpages: add accordingly. 
  * debian/rules: updated to install the new binary package.
  * debian/copyright: updated 
  * config/*: add to give a custom experience to grotta's user
  * debian/grotta-install: updated accordingly

 -- Andrea Colangelo <warp10@debian.org>  Mon, 26 Oct 2015 10:43:22 +0100

buc (4.2ubuntu0) precise; urgency=medium

  * shut.buc: password hash update.
  * shut.*: update copyright year.
  * debian/copyright: updated accordingly.

 -- Andrea Colangelo <warp10@debian.org>  Thu, 09 Apr 2015 19:11:23 +0200

buc (4.1ubuntu0) precise; urgency=low

  * shut.*: fix a f***ing bug that avoids logout() being called.

 -- Andrea Colangelo <warp10@ubuntu.com>  Tue, 17 Sep 2013 22:23:10 +0200

buc (4ubuntu0) precise; urgency=low

  * Removal of "Asino chi non legge" stuff:
    + asino_chi_non_legge.png: drop.
    + modello_iscrizioni_asino_chi_non_legge.doc: drop.
    + auto.sh.*: update accordingly.
    + debian/*.install: update accordingly
  * shut.*: fix timeout not properly working.
  * Install auto.sh in /usr/share/buc:
    + debian/*.install: update accordingly.
    + debian/rules: update accordingly.
    + guest-session-auto: update accordingly.
  * Add new package buc-common containing shared files:
    + debian/control: add buc-common stanza and modified others accordingly.
    + debian/*.install: install common files in buc-common.
    + debian/*.post{inst,rm}: moved to buc-common.
    * debian/*.docs: moved to buc-common.

 -- Andrea Colangelo <warp10@ubuntu.com>  Thu, 12 Sep 2013 12:09:12 +0200

buc (3.1ubuntu0~ppa1) precise; urgency=low

  * shut: split into shut.buc and shut.biblio.
  * shut.1: split into shut.buc.1 and shut.biblio.1 accordingly.
  * auto.sh.*: launch shut.* accordingly.
  * auto.sh.biblio: re-enable dialog window, add RADIUS disconnect and shut.
  * debian/*.install: install the right version of shut for each package.
  * debian/manpages: split in two files and update.
  * debian/docs: split in two files.
  * debian/post{install,rm}: split in two files.
  * modello_iscrizioni_asino_chi_non_legge.doc:
    + add, temporary files to promote the reading marathon.
    + debian/*.install: modify to install the file
    + auto.sh.*: modify to cp the file to Desktop
  * asino_chi_non_legge.png:
    + add, temporary wallpaper to promote the reading marathon.
    + debian/*.install: modify to install the file
    + auto.sh.: modify to set as wallpaper.

 -- Andrea Colangelo <warp10@ubuntu.com>  Wed, 10 Jul 2013 10:11:29 +0200

buc (3.0ubuntu0~ppa1) precise; urgency=low

  * debian/control:
    + change Vcs-*, code is now hosted on github (see you, bzr!)
    + add biblio binary package;
    + buc binary stanza modified accordingly.
  * debian/*.install: update accordingly.
  * debian/biblio.manpages: add accordingly.
  * {un,}install.sh: drop, useless.
  * auto.sh*: split auto.sh in auto.sh.buc and auto.sh.biblio.
  * debian/rules: override dh_install to properly mangle auto.sh*.

 -- Andrea Colangelo <warp10@ubuntu.com>  Thu, 13 Jun 2013 18:30:47 +0200

buc (2.1ubuntu0~ppa1) precise; urgency=low

  * auto.sh: start x11vnc
  * debian/control: add some deps

 -- Andrea Colangelo <warp10@ubuntu.com>  Wed, 12 Jun 2013 17:23:41 +0200

buc (2.0ubuntu0~ppa1) precise; urgency=low

  * shut: add window to authorize user and start session.  
  * auto.sh:
    + Do not create the firefox icon anymore.
    + Do not use zenity to create the warning window. 

 -- Andrea Colangelo <warp10@ubuntu.com>  Wed, 12 Jun 2013 17:23:28 +0200

buc (1.0ubuntu0~ppa0) precise; urgency=low

  * home.html: improved, based on firefox default page.
  * images/*: added accordingly.

 -- Andrea Colangelo <warp10@ubuntu.com>  Thu, 14 Feb 2013 20:19:15 +0100

buc (0.2ubuntu0~ppa1) precise; urgency=low

  * prefs.sh: drop code to handle locales.
  * shut:
    + remaining time now shown next to the icon;
    + changed order of options in menu;
    + PEP8'ed.
  * auto.sh: info dialog improved.
  * debian/post{inst,rm}: disable xterm and uxterm too.

 -- Andrea Colangelo <warp10@ubuntu.com>  Thu, 14 Feb 2013 12:00:17 +0100

buc (0.1ubuntu0~ppa1) precise; urgency=low

  * Initial Release.

 -- Andrea Colangelo <warp10@ubuntu.com>  Mon, 11 Feb 2013 11:40:45 +0100


buc/shut.py
#!/usr/bin/python

# BUC Customizer
#
# Copyright (C) 2012-2013 Andrea Colangelo
#
# Authors: Andrea Colangelo <warp10@ubuntu.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; version 3 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, see: http://www.gnu.org/licenses/gpl-3.0.txt


from gi.repository import AppIndicator3 as appindicator
from gi.repository import Gtk
from gi.repository import GLib
from gi.repository import Notify
from os import system
from time import time, sleep


class Shut():
    def __init__(self):
        self.icon = "/usr/share/icons/hicolor/scalable/apps/shut.svg"

        self.start = time()

        self.show_notice_30()
        GLib.timeout_add_seconds(1200, self.show_notice_10)
        GLib.timeout_add_seconds(1740, self.show_notice_2)
        GLib.timeout_add_seconds(1800, self.show_notice_0)

        self.load_indicator()

    def load_indicator(self):
        self.ind = appindicator.Indicator.new("shut",
                             self.icon,
                             appindicator.IndicatorCategory.APPLICATION_STATUS)
        self.ind.set_status(appindicator.IndicatorStatus.ACTIVE)

        menu = Gtk.Menu()

        buf = "Termina sessione"
        menu_items = Gtk.MenuItem(buf)
        menu.append(menu_items)
        menu_items.connect("activate", self.logout, buf)
        menu_items.show()

        buf = "Minuti disponibili"
        menu_items = Gtk.MenuItem(buf)
        menu.append(menu_items)
        menu_items.connect("activate", self.remaining, buf)
        menu_items.show()

        menu_items = Gtk.SeparatorMenuItem()
        menu.append(menu_items)
        menu_items.show()

        buf = "Info"
        menu_items = Gtk.MenuItem(buf)
        menu.append(menu_items)
        menu_items.connect("activate", self.about, buf)
        menu_items.show()

        #buf = "Chiudi"
        #menu_items = Gtk.MenuItem(buf)
        #menu.append(menu_items)
        #menu_items.connect("activate", Gtk.main_quit, buf)
        #menu_items.show()

        menu.show()

        self.ind.set_menu(menu)

        Gtk.main()

    def show_notice_30(self):
        notice = "Questa sessione sara' terminata automaticamente tra 30 minuti."
        self.show_notice(notice)
        return False

    def show_notice_10(self):
        notice = "Questa sessione sara' terminata automaticamente tra 10 minuti."
        self.show_notice(notice)
        return False

    def show_notice_2(self):
        self.icon = "/usr/share/icons/hicolor/scalable/apps/shutr_red.svg"
        self.ind.set_icon_full(self.icon, "shut")
        notice = "Questa sessione sara' terminata automaticamente tra 2 minuti. \
                  Concludi il lavoro in corso e salva i tuoi documenti. "
        self.show_notice(notice)
        return False

    def show_notice_0(self):
        notice = "Chiusura della sessione in corso!"
        self.show_notice(notice)
        sleep(5)
        self.logout()

    def remaining(self, *args):
        elapsed_seconds = int(time() - self.start)
        remaining_minutes = (1800 - elapsed_seconds) / 60
        notice = "Restano " + str(remaining_minutes) + " minuti disponibili"
        self.show_notice(notice, header="Tempo disponibile")

    def show_notice(self, notice, header="Avviso: chiusura sessione"):
        icon = self.icon
        Notify.init('Shut')
        n = Notify.Notification.new(header, "\n" + notice, icon)
        n.show()

    def logout(self, *args):
        cmd = "dbus-send --session --type=method_call --dest=org.gnome.SessionManager /org/gnome/SessionManager org.gnome.SessionManager.Logout uint32:1"
        system(cmd)

    def about(self, *args):
        a = Gtk.AboutDialog()
        a.set_program_name("Shut")
        a.set_comments("Tool to Logout User")
        a.set_version("0.1")
        a.set_copyright("Copyright (C) 2012-2013 Andrea Colangelo")
        a.set_license_type(Gtk.License.GPL_3_0)
        a.set_website("http://www.linuxfm.org")
        a.set_artists(["Marco Alici", "Copyright (C) 2013 Marco Alici", "Released under the terms of CC-BY-SA 3.0, see:", "http://creativecommons.org/licenses/by-sa/3.0/"])
        a.run()
        a.destroy()

if __name__ == '__main__':
    shut = Shut()
